### ISSUE TITLE
Fatal error: exception Xs_protocol.Error("EACCES [read]")

### ISSUE THREAD

**Issue by talex5** [talex5/qubes-mirage-firewall#2]
Have got this a couple of times:

    2016-01-06 22:06.57: WRN [firewall] Failed to parse frame
    Fatal error: exception Xs_protocol.Error("EACCES [read]")
    Raised at file "bytes.ml", line 220, characters 25-34
    Called from file "core/xs_protocol.ml", line 77, characters 14-30
    Mirage exiting with status 2
    Do_exit called!

Will add a better error to XenStore and hope it happens again...

**Comment by talex5** [comment]
Fatal error: exception Xs_protocol.Error("EACCES [read:\"/local/domain/212/device/vif/0/state\\000\"]")

**Comment by talex5** [comment]
Better error from new logging system. Happened after starting a client that exited immediately:

    --- Dumping log ring buffer ---
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.write "backend/vif/6/0/feature-rx-flip"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.write "backend/vif/6/0/feature-rx-notify"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.write "backend/vif/6/0/feature-smart-poll"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.write "backend/vif/6/0/state"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.directory "backend/vif"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.read "backend/vif/6/0/frontend-id"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.directory "backend/vif"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.read "backend/vif/6/0/frontend"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.read "/local/domain/6/device/vif/0/backend-id"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.read "/local/domain/6/device/vif/0/backend"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.read "backend/vif/6/0/frontend"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.read "/local/domain/6/device/vif/0/state"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.read "/local/domain/6/device/vif/0/state"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.read "/local/domain/6/device/vif/0/state"
    2016-01-13 10:23.51: INF [qubes.db] got update: "/qubes-iptables-header" = "# Generated by Qubes Core on Wed Jan 13 10:23:51 2016\n*filter\n:INPUT DROP [0:0]\n:FORWARD DROP [0:0]\n:OUTPUT ACCEPT [0:0]\n-A INPUT -i vif+ -p udp -m udp --dport 68 -j DROP\n-A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT\n-A INPUT -p icmp -j ACCEPT\n-A INPUT -i lo -j ACCEPT\n-A INPUT -j REJECT --reject-with icmp-host-prohibited\n-A FORWARD -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT\n-A FORWARD -i vif0.0 -j ACCEPT\n-A FORWARD -i vif+ -o vif+ -j DROP\nCOMMIT\n"
    2016-01-13 10:23.51: INF [qubes.db] got update: "/qubes-iptables-domainrules/6" = "*filter\n-A FORWARD -s 10.137.4.11 -d 80 -p tcp --dport 80 -j REJECT --reject-with icmp-host-prohibited\n-A FORWARD -s 10.137.4.11 -p udp -d 10.137.1.1 --dport 53 -j ACCEPT\n-A FORWARD -s 10.137.4.11 -p udp -d 10.137.1.254 --dport 53 -j ACCEPT\n-A FORWARD -s 10.137.4.11 -p tcp -d 10.137.1.1 --dport 53 -j ACCEPT\n-A FORWARD -s 10.137.4.11 -p tcp -d 10.137.1.254 --dport 53 -j ACCEPT\n-A FORWARD -s 10.137.4.11 -p icmp -j ACCEPT\n-A FORWARD -s 10.137.4.11 -p tcp -d 10.137.255.254 --dport 8082 -j DROP\n-A FORWARD -s 10.137.4.11 -j ACCEPT\nCOMMIT\n"
    2016-01-13 10:23.51: INF [qubes.db] got update: "/qubes-netvm-domid" = "1"
    2016-01-13 10:23.51: INF [qubes.db] got update: "/qubes-iptables" = "reload"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.read "/local/domain/6/device/vif/0/state"
    2016-01-13 10:23.51: DBG [xenstore.client] Xs.directory "backend/vif"
    --- End dump ---
    Fatal error: exception Xs_protocol.Error("EACCES [read:\"/local/domain/6/device/vif/0/state\\000\"]")
    Raised at file "bytes.ml", line 220, characters 25-34
    Called from file "core/xs_protocol.ml", line 77, characters 14-30
    Mirage exiting with status 2
    Do_exit called!

**Comment by talex5** [comment]
This was due to a buggy patch I made to XenStore (so it won't affect other people). The upstream XenStore version of `wait` would silently ignore the error. My patch caused it to abort the unikernel. I've now updated the patch to fail just the wait thread, allowing the firewall to recover.

Now you just get a warning logged:

    2016-01-13 10:50.01: WRN [net] Error connecting client domain 10: Xs_protocol.Error("EACCES [read:\"/local/domain/10/device/vif/0/state\\000\"]")

(XenStrore PR: https://github.com/mirage/ocaml-xenstore/pull/13)
